internal/telemetry/telemetry_test.go

```diff
@@ -36,3 +41,116 @@ func TestProductChange(t *testing.T) {
 	var productsPayload *Products = secondBody.Payload.(*Products)
 	assert.Equal(t, productsPayload.Profiler.Enabled, true)
 }
+
+// mockServer initializes a server that expects a strict amount of telemetry events. It saves these
+// events in a slice until the expected number of events is reached.
+// the `telemetry` argument accepts a function that should run some sequence of telemetry events.
+// the `expectedHits` argument specifies the number of telemetry events the server should expect.
+func mockServer(ctx context.Context, t *testing.T, expectedHits int, telemetry func(), exclude ...RequestType) (waitForEvents func() []string, cleanup func()) {
+	messages := make([]string, expectedHits)
+	hits := 0
+	done := make(chan struct{})
+	mu := sync.Mutex{}
+	excludeEvent := make(map[RequestType]struct{})
+	for _, event := range exclude {
+		excludeEvent[event] = struct{}{}
+	}
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path != "/telemetry/proxy/api/v2/apmtelemetry" {
+			return
+		}
+		rType := RequestType(r.Header.Get("DD-Telemetry-Request-Type"))
+		if _, ok := excludeEvent[rType]; ok {
+			return
+		}
+		mu.Lock()
+		defer mu.Unlock()
+		if hits == expectedHits {
+			t.Fatalf("too many telemetry messages (expected %d)", expectedHits)
+		}
+		messages[hits] = string(rType)
+		hits++
+		if hits == expectedHits {
+			done <- struct{}{}
+		}
+	}))
+	GlobalClient.ApplyOps(WithURL(false, server.URL))
+
+	return func() []string {
+			telemetry()
+			select {
+			case <-ctx.Done():
+				t.Fatal("TestProductStart timed out")
+			case <-done:
+			}
+			return messages
+		}, func() {
+			server.Close()
+			GlobalClient.Stop()
+		}
+}
+
+func TestProductStart(t *testing.T) {
+	// this test is meant to ensure that a given sequence of ProductStart/ProductStop calls
+	// emits the expected telemetry events.
+	t.Setenv("DD_TELEMETRY_HEARTBEAT_INTERVAL", "1")
+	t.Setenv("DD_TRACE_STARTUP_LOGS", "0")
+	tests := []struct {
+		name           string
+		wantedMessages []string
+		telemetry      func()
+	}{
+		{
+			name:           "tracer start, profiler start with config",
+			wantedMessages: []string{"app-started", "app-dependencies-loaded", "app-client-configuration-change", "app-product-change"},
```

Comment:
Aren't these strings defined as constants already? Can we use those constants here (even though they're a different type?)